finetune/mtl.py

```diff
@@ -0,0 +1,173 @@
+                val_size=val_size
+            )
+            self.dataset_size_ += self.config.dataset_size
+            frequencies.append(self.config.dataset_size)
+            (val_func, input_func, val_sizes[task_name], val_intervals[task_name]) = task_tuple
+            task_id = self.config.task_name_to_id[task_name]
+
+            input_func_normalised, val_func_normalised = get_input_fns(task_id, input_func, val_func)
+            input_funcs.append(input_func_normalised)
+            val_funcs[task_name] = val_func_normalised
+            val_funcs[task_name + "_train"] = get_train_eval_dataset(input_func_normalised, val_sizes[task_name])
+
+        sum_frequencies = sum(frequencies)
+        weights = [float(w) / sum_frequencies for w in frequencies]
+        train_dataset = lambda: tf.data.experimental.sample_from_datasets([f() for f in input_funcs], weights)
+
+        self.config.task_input_pipelines = input_pipelines
+        self.config.dataset_size = self.dataset_size_
+        return val_funcs, train_dataset, val_sizes, val_intervals
+
+    def _target_encoder(self):
+        raise FinetuneError("This should never be used??")
+
+
+def get_loss_logits_fn(task, featurizer_state, config, targets_i, train, reuse, task_id_i):
+    def loss_logits():
+        with tf.variable_scope("target_model_{}".format(task)):
+            target_model_out = config.tasks[task]._target_model(
+                config=config,
+                featurizer_state=featurizer_state,
+                targets=targets_i,
+                n_outputs=config.task_input_pipelines[task].target_dim,
+                train=train,
+                reuse=reuse
+            )
+            logits = target_model_out["logits"]
+            logits.set_shape(None)
+            return target_model_out["losses"], logits
+
+    return tf.equal(task_id_i, config.task_name_to_id[task]), loss_logits
+
+
+class MultiTask(BaseModel):
+
+    def __init__(self, tasks, **kwargs):
+        super().__init__(**kwargs)
+        self.config.tasks = tasks
+        self.config.task_name_to_id = dict(zip(self.config.tasks.keys(), range(len(self.config.tasks))))
+
+    def _get_input_pipeline(self):
+        return MultiTaskPipeline(self.config)
+
+    def featurize(self, X):
+        return super().featurize(X)
+
+    def cached_predict(self):
+        raise FinetuneError("cached_predict is not supported yet for MTL")
+
+    def predict(self, X):
+        predictions = {}
+        for name, ModelClass in self.config.tasks.items():
+            if name not in X:
+                continue
+            pred_model = ModelClass()
+            pred_model.input_pipeline = self.config.task_input_pipelines[name]
+            pred_model.saver.variables = {
+                k.replace("/target_model_{}".format(name), ""): v for k, v in self.saver.variables.items()
+            }
+            predictions[name] = pred_model.predict(X[name])
+        return predictions
+
+    def predict_proba(self, X):
```

Comment:
This should follow the same format, right?  Maybe w/ a fallback if this is unsupported by the task model?

Reply:
Works for me